Blog : Building a distributed tracing pipeline

kuberkaul opened the pull request:
Signed-off-by kuber.kaul@dowjones.com

### Description
This a blog trying to document Building a distributed tracing pipeline with open telemetry collector, data prepper, Jaeger and OpenSearch trace analytics
 
### Issues Resolved
- https://github.com/opensearch-project/project-website/issues/444

### Check List
- [X ] Commits are signed per the DCO using --signoff


By submitting this pull request, I confirm that my contribution is made under the terms of the BSD-3-Clause License.

kuberkaul commented:
Hope I did this right and the blog is useful. Let me know if I need to make any correction needed.

kuberkaul commented:
Fixed the sign off.

stockholmux reviewed (changes requested):
This is an **awesome** blog post. I'm really excited to get it on the site!

There are a few things that need to be changed - mostly small edits.

I'm also seeing that you're missing a bio (and an author photo, if possible) - we'll need that to publish. Here is my author bio as an example: https://github.com/opensearch-project/project-website/blob/main/_authors/kyledvs.markdown?plain=1

I'm going to ask a few folks from the Data Prepper team to review as well.

kuberkaul commented on _posts/2021-12-02-distributed-tracing-pipeline-with-opentelemetry.md:
```diff
@@ -0,0 +1,895 @@
+
+- **OpenTelemetry** for creation, propagation, collection, processing, and exporting of trace data.
+
+- **AWS OpenSearch** (formerly ElasticSearch) as the sink for the traces
```
✔️

kuberkaul commented on _posts/2021-12-02-distributed-tracing-pipeline-with-opentelemetry.md:
```diff
@@ -0,0 +1,895 @@
+
+[Jaeger](https://www.jaegertracing.io/) tracks and measures requests and transactions by analyzing end-to-end data from service call chains, making it easier to understand latency issues in microservice architectures. It's exclusively meant for distributed tracing with a simple web UI to query traces and spans across services. Jaeger enables views of individual traces in a waterfall-style graph o ...
+
+Data Prepper is a new component of OpenSearch that receives trace data from the OpenTelemetry collector, aggregates, transforms, and normalizes it for analysis and visualization in Kibana.
```
✔️

kuberkaul commented on _posts/2021-12-02-distributed-tracing-pipeline-with-opentelemetry.md:
```diff
@@ -0,0 +1,895 @@
+# Building a distributed tracing pipeline with open telemetry collector, data prepper, and OpenSearch trace analytics
```
✔️

kuberkaul commented on _posts/2021-12-02-distributed-tracing-pipeline-with-opentelemetry.md:
```diff
@@ -0,0 +1,895 @@
+
+The landscape for observability tools continues to grow as we speak. In particular, when it comes to metrics, error logging, and distributed traces. These can provide valuable information to optimize service performance and troubleshoot issues to make a service more reliable. With this in mind, it makes sense to create a distributed tracing pipeline to ingest, process, and visualize tracing data  ...
+
+At [Dow Jones](#TODO), we started on a similar journey as we continue to move our applications and microservices to our service mesh based on [EKS](https://docs.aws.amazon.com/whitepapers/latest/overview-deployment-options/amazon-elastic-kubernetes-service.html) and [istio](https://istio.io/). Our requirements formulated from the generated service mesh telemetry data was to have a distributed tra ...
```
✔️

kuberkaul commented on _posts/2021-12-02-distributed-tracing-pipeline-with-opentelemetry.md:
```diff
@@ -0,0 +1,895 @@
+
+Data Prepper is a new component of OpenSearch that receives trace data from the OpenTelemetry collector, aggregates, transforms, and normalizes it for analysis and visualization in Kibana.
+
+- **Trace Analytics Plugin** for opensearch to visualize and query traces
```
✔️

kuberkaul commented on _posts/2021-12-02-distributed-tracing-pipeline-with-opentelemetry.md:
```diff
@@ -0,0 +1,895 @@
+- [W3C Correlation-Context HTTP Propagator](https://w3c.github.io/correlation-context/)
+- [B3 Zipkin HTTP Propagator](https://github.com/openzipkin/b3-propagation)
+
+This works well for our service mesh PaaS as Istio leverages Envoy’s distributed tracing feature to provide tracing integration out of the box. Specifically, Istio provides options to install various tracing backend and configure proxies to send trace spans to them automatically. It requires an application to propagate the [B3 Zipkin HTTP Propagator](https://github.com/openzipkin/b3-propagation)  ...
```
✔️  fixed wording.

kuberkaul commented on _posts/2021-12-02-distributed-tracing-pipeline-with-opentelemetry.md:
```diff
@@ -0,0 +1,895 @@
+
+We will be deploying OpenTelemetry agents as daemonset to receive, process, and export trace from every EKS worker node. Agent is capable of receiving telemetry data (push and pull based) as well as enhancing telemetry data with metadata such as custom tags or infrastructure information. In addition, the agent can offload responsibilities that client instrumentation would otherwise need to handle ...
+
+The agent can be deployed either as a daemonset or as a sidecar in a Kubernetes cluster. This step may be skipped (not recommended), if you are creating this pipeline in a test environment and would rather send traces straight to the open telemetry collector, which is running as a deployment (horizontally scalable)
```
✔️

kuberkaul commented on _posts/2021-12-02-distributed-tracing-pipeline-with-opentelemetry.md:
```diff
@@ -0,0 +1,895 @@
+
+OpenTelemetry (OTEL) is formed by the merging of OpenTracing and OpenCensus. It is a CNCF incubating project and the second most active in terms of contributions. OTEL, since its inception, aimed to offer a single set of APIs and libraries that standardize how you collect and transfer telemetry data.
+
+[Traces was the first component to reach stable status for the open telemetry project](https://opentelemetry.io/status/). Since then, almost all trace components but collector has reached stable maturity level, metrics are well on their way, and initial work on logs has also started. At Dow Jones, we launched a distributed tracing pipeline based on the OpenTelemetry project. I will be going over  ...
```
✔️

kuberkaul commented on _posts/2021-12-02-distributed-tracing-pipeline-with-opentelemetry.md:
```diff
@@ -0,0 +1,895 @@
+
+- **latency**: Sample based on the duration of the trace. The duration is determined by looking at the earliest start time and latest end time, without taking into consideration what happened in between.
+- **probabilistic**: Sample a percentage of traces.
+- **status_code**: Sample based upon the status code (OK, ERROR or UNSET)
```
✔️

kuberkaul commented on _posts/2021-12-02-distributed-tracing-pipeline-with-opentelemetry.md:
```diff
@@ -0,0 +1,895 @@
+
+- **&#8594;** In the agent config, add a span attribute for all traces with a key/value pair (e.g - retain_span/false) using attribute processor.
+- **&#8594;** In the agent config, add another attribute processor now to override that value to false for cherrypicked services that we don't want to be sampled.
+- **&#8594;** In the gateway config, use the string attribute filter on tail based sampling with the **key: retain_span** and **value: true**. All traces without *retain_span: true* attribute will not be sampled.
```
✔️

kuberkaul commented on _posts/2021-12-02-distributed-tracing-pipeline-with-opentelemetry.md:
```diff
@@ -0,0 +1,895 @@
+
+### Sink : OpenSearch
+
+As you must have noticed in the architecture, we utilize OpenSearch to ship to and store traces. [Opensearch](https://aws.amazon.com/blogs/opensource/introducing-opensearch/) is a community-driven, open source fork of Elasticsearch and Kibana. This project includes OpenSearch (derived from Elasticsearch 7.10.2) and OpenSearch Dashboards (derived from Kibana 7.10.2). Additionally, the OpenSearch p ...
```
✔️

kuberkaul commented on _posts/2021-12-02-distributed-tracing-pipeline-with-opentelemetry.md:
```diff
@@ -0,0 +1,895 @@
+
+**&#8594;** The collector processes/samples the traces and export them to backends which in our case are data prepper and jaeger collector. These backends transform and export the data to opensearch.
+
+**&#8594;** Trace Analytics plugin for Kibana and Jaeger visualizes trace data from opensearch and allows querying on it.
```
✔️

kuberkaul commented on _posts/2021-12-02-distributed-tracing-pipeline-with-opentelemetry.md:
```diff
@@ -0,0 +1,895 @@
+
+### Formatting and Exporting Traces
+
+We now have distributed traces created, context propagated, and sampled using tail based sampling. We need to format the trace data in a way that our query engine can analyze it. At Dow Jones we have teams that rely on Jaeger and Kibana Trace Analytics to query and visualize this data. With OpenTelemetry, we can push the data to multiple backends from the same collector. At present, there is one  ...
```
✔️

kuberkaul commented on _posts/2021-12-02-distributed-tracing-pipeline-with-opentelemetry.md:
```diff
@@ -0,0 +1,895 @@
+
+- **Trace Analytics feature in OpenSearch/Kibana**
+
+[Amazon Elasticsearch Service released Trace Analytics](https://aws.amazon.com/blogs/big-data/getting-started-with-trace-analytics-in-amazon-elasticsearch-service/), a new feature for distributed tracing that enables developers and operators to troubleshoot performance and availability issues in their distributed applications, giving them end-to-end insights not possible with traditional methods  ...
```
✔️

kuberkaul commented on _posts/2021-12-02-distributed-tracing-pipeline-with-opentelemetry.md:
```diff
@@ -0,0 +1,895 @@
+
+We now have the distributed tracing pipeline. It is time to start an application that generates traces and exports them to agents. Depending on your application code and whether you want to do manual or automatic instrumentation, the [Open Telementry Instrumentation Docs](https://opentelemetry.io/docs/) should get you started.
+
+In this setup, OTEL agents are running as daemonset. We will be exporting traces from the service to the agent on the same host(worker-node). To get the host IP, we can set the HOST_IP environment variable via the [Kubernetes downwards API](https://kubernetes.io/docs/tasks/inject-data-application/downward-api-volume-expose-pod-information/#capabilities-of-the-downward-api). It can be referenced i ...
```
✔️

kuberkaul commented on _posts/2021-12-02-distributed-tracing-pipeline-with-opentelemetry.md:
```diff
@@ -0,0 +1,895 @@
+
+We built a scalable distributed tracing pipeline based on the OpenTelemetry project running in an EKS cluster. One of the driving forces behind our move to OpenTelemetry is to consolidate to a single binary to ingest, process, and export, not just traces but metrics and logs as well (telemetry data). Telemetry data can be exported to multiple backends by creating [pipelines](https://github.com/op ...
+
+Data Prepper 1.2 (December 2021) release is going to provide users the ability to send logs from Fluent Bit to OpenSearch or Amazon OpenSearch Service and use Grok to enhance the logs. Logs can then be correlated to traces coming from the OTEL collectors to further enhance deep diving into your service problems using OpenSearch Dashboards. This should elevate the experience of deep diving into yo ...
```
✔️

kuberkaul commented on _posts/2021-12-02-distributed-tracing-pipeline-with-opentelemetry.md:
```diff
@@ -0,0 +1,895 @@
+# Building a distributed tracing pipeline with open telemetry collector, data prepper, and OpenSearch trace analytics
```
✔️  used is as is, just changed name to `Kuber Kaul`

kuberkaul commented on _posts/2021-12-02-distributed-tracing-pipeline-with-opentelemetry.md:
```diff
@@ -0,0 +1,895 @@
+
+K8s deployments are highly elastic and can be automatically scaled up and down via Horizontal Pod Autoscale depending on the amount of traces flowing through the pipeline.
+
+<img src="../assets/media/blog-images/2021-12-02-distributed-tracing-pipeline-with-opentelemetry/tracing_step2.png" align="center">
```
✔️

kuberkaul commented on _posts/2021-12-02-distributed-tracing-pipeline-with-opentelemetry.md:
```diff
@@ -0,0 +1,895 @@
+
+OpenSearch makes an excellent choice for storing and searching trace data, along with other observability data due to its fast search capabilities and horizontal scalability.
+
+<img src="../assets/media/blog-images/2021-12-02-distributed-tracing-pipeline-with-opentelemetry/tracing_step3.png" align="center" >
```
✔️

kuberkaul commented on _posts/2021-12-02-distributed-tracing-pipeline-with-opentelemetry.md:
```diff
@@ -0,0 +1,895 @@
+Now is the part where you actually get to use the pipeline and visualize these traces to run queries, build dashboards, setup alerting. While Trace Analytics OpenSearch Dashboards plugin is a managed plugin that comes with OpenSearch, we host our own [Jaeger Frontend/UI](https://www.jaegertracing.io/docs/1.28/frontend-ui/) with a simple jaeger UI to query for distributed traces with Traces View
+and Traces Detail View. We deploy Jaeger Query as a deployment within the EKS cluster. Jaeger query can be deployed as a deployment with a horizontal pod autoscaler.
+
+<img src="../assets/media/blog-images/2021-12-02-distributed-tracing-pipeline-with-opentelemetry/tracing_step4.png" align="center" >
```
✔️

kuberkaul commented on _posts/2021-12-02-distributed-tracing-pipeline-with-opentelemetry.md:
```diff
@@ -0,0 +1,895 @@
+
+Once you have traces flowing through your pipeline, you should see indexes being created in OpenSearch under the otel-v1\* index prefix. As this is created, trace analytics plugin within OpenSearch dashboards should now have visualization for your traces, inclusive of service map, error/throughput graphs, and waterfall trace view for your services.
+
+<img src="../assets/media/blog-images/2021-12-02-distributed-tracing-pipeline-with-opentelemetry/kibana.png" align="center" height="50%"  width="50%">
```
✔️

kuberkaul commented on _posts/2021-12-02-distributed-tracing-pipeline-with-opentelemetry.md:
```diff
@@ -0,0 +1,895 @@
+
+**&#8594;** Trace Analytics plugin for Kibana and Jaeger visualizes trace data from opensearch and allows querying on it.
+
+<img src="../assets/media/blog-images/2021-12-02-distributed-tracing-pipeline-with-opentelemetry/tracing_step4.png" align="center" >
```
✔️  I changed the name of the image to `opensearch` and the other image with kibana logo was not actually being used so I just removed it

kuberkaul commented on _posts/2021-12-02-distributed-tracing-pipeline-with-opentelemetry.md:
```diff
@@ -0,0 +1,895 @@
+
+</details>
+
+- **Trace Analytics feature in OpenSearch/Kibana**
```
✔️

kuberkaul commented:
All changes done as per feedback !

kuberkaul commented:
btw I just noticed this :

> I'm also seeing that you're missing a bio (and an author photo, if possible) - we'll need that to publish. Here is my author bio as an example: https://github.com/opensearch-project/project-website/blob/main/_authors/kyledvs.markdown?plain=1

I'll work on this. Also I have been working with Rajiv Taori (https://opensearch.org/authors/rajtaori/) and his team, so I am going to ask him for a review as well, cant seem to @ him here.

kuberkaul commented:
author and image added now. ✔️

dlvenable commented on _posts/2021-12-02-distributed-tracing-pipeline-with-opentelemetry.md:
```diff
@@ -0,0 +1,904 @@
+
+OpenTelemetry (OTEL) is formed by the merging of OpenTracing and OpenCensus. It is a CNCF incubating project and the second most active in terms of contributions. OTEL, since its inception, aimed to offer a single set of APIs and libraries that standardize how you collect and transfer telemetry data.
+
+[Traces was the first component to reach stable status for the OpenTelemetry project](https://opentelemetry.io/status/). Since then, almost all trace components but collector has reached stable maturity level, metrics are well on their way, and initial work on logs has also started. At Dow Jones, we launched a distributed tracing pipeline based on the OpenTelemetry project. I will be going over a ...
```
> Traces was the first component to reach stable status

Perhaps: "Tracing was the first component to reach stable status"

kuberkaul replied:
✔️

dlvenable reviewed (changes requested):
Thank you Kuber for making this post! I found it very helpful to see how your team is using these different components for observability.  I also like that you included configuration snippets in collapsed sections.

I made a few suggestions throughout the document.

kuberkaul commented on _posts/2021-12-02-distributed-tracing-pipeline-with-opentelemetry.md:
```diff
@@ -0,0 +1,895 @@
+
+Over the past few years, the importance of observability when developing and managing applications has spiked with the spotlight firmly on micro-services, service mesh. Distributed services can be unpredictable. Despite our best efforts, failures and performance bottlenecks in such systems are inevitable and difficult to isolate. In such an environment, having deep visibility into the behavior of ...
+
+The landscape for observability tools continues to grow as we speak. In particular, when it comes to metrics, error logging, and distributed traces. These can provide valuable information to optimize service performance and troubleshoot issues to make a service more reliable. With this in mind, it makes sense to create a distributed tracing pipeline to ingest, process, and visualize tracing data  ...
```
✔️

kuberkaul commented on _posts/2021-12-02-distributed-tracing-pipeline-with-opentelemetry.md:
```diff
@@ -0,0 +1,895 @@
+
+**&#8594;** Trace Analytics plugin for Kibana and Jaeger visualizes trace data from opensearch and allows querying on it.
+
+<img src="../assets/media/blog-images/2021-12-02-distributed-tracing-pipeline-with-opentelemetry/tracing_step4.png" align="center" >
```
✔️

kuberkaul commented on _posts/2021-12-02-distributed-tracing-pipeline-with-opentelemetry.md:
```diff
@@ -0,0 +1,895 @@
+
+**&#8594;** OpenTelemetry agents process/batches and send traces from microservices to the openTelemetry gateway collector.
+
+**&#8594;** The collector processes/samples the traces and export them to backends which in our case are data prepper and jaeger collector. These backends transform and export the data to opensearch.
```
✔️

kuberkaul commented on _posts/2021-12-02-distributed-tracing-pipeline-with-opentelemetry.md:
```diff
@@ -0,0 +1,904 @@
+
+**&#8594;** OpenTelemetry agents process/batches and send traces from microservices to the openTelemetry gateway collector.
+
+**&#8594;** The collector processes/samples the traces and export them to backends which in our case are data prepper and jaeger collector. These backends transform and export the data to OpenSearch.
```
✔️

kuberkaul commented on _posts/2021-12-02-distributed-tracing-pipeline-with-opentelemetry.md:
```diff
@@ -0,0 +1,904 @@
+
+- **Data Prepper**
+
+Similar to Jaeger, to vizualize distributed traces through Trace Analytics in OpenSearch, we need to transform these traces for OpenSearch. Data Prepper is a key component in providing Trace Analytics feature in OpenSearch. Data Prepper is a last mile server-side component collecting telemetry data from AWS Distro OpenTelemetry collector or OpenTelemetry collector and transforms it for OpenSearch ...
```
✔️

kuberkaul commented on _posts/2021-12-02-distributed-tracing-pipeline-with-opentelemetry.md:
```diff
@@ -0,0 +1,904 @@
+
+### Step 2 : OpenTelemetry Agents
+
+We will be deploying OpenTelemetry agents as daemonset to receive, process, and export trace from every EKS worker node. Agent is capable of receiving telemetry data (push and pull based) as well as enhancing telemetry data with metadata such as custom tags or infrastructure information. In addition, the agent can offload responsibilities that client instrumentation would otherwise need to handle ...
```
✔️

kuberkaul commented on _posts/2021-12-02-distributed-tracing-pipeline-with-opentelemetry.md:
```diff
@@ -0,0 +1,904 @@
+
+### Formatting and Exporting Traces
+
+We now have distributed traces created, context propagated, and sampled using tail based sampling. We need to format the trace data in a way that our query engine can analyze it. At Dow Jones, we have teams that rely on Jaeger and OpenSearch Dashboards to query and visualize this data. With OpenTelemetry, we can push the data to multiple backends from the same collector. At present, there is one  ...
```
✔️

kuberkaul commented on _posts/2021-12-02-distributed-tracing-pipeline-with-opentelemetry.md:
```diff
@@ -0,0 +1,904 @@
+
+![tracing_step3](/assets/media/blog-images/2021-12-02-distributed-tracing-pipeline-with-opentelemetry/tracing_step3.png){: .img-fluid}
+
+Jaeger collector and Data Prepper can be configured to ship traces to the same OpenSearch. Both deployments create their indexes with a mutually exclusive prefix for name. OpenSearch can be configured with index patterns to create seperate views:
```
✔️

kuberkaul commented on _posts/2021-12-02-distributed-tracing-pipeline-with-opentelemetry.md:
```diff
@@ -0,0 +1,904 @@
+
+- **Trace Analytics feature in OpenSearch/OpenSearch Dashboards**
+
+Trace Analytics for Opensearch Dashboards can be levaraged without having to re-instrument applications. Once you have traces flowing through your pipeline, you should see indexes being created in OpenSearch under the `otel-v1\*` index prefix. As this is created, trace analytics plugin within OpenSearch dashboards should now have visualization for your traces, inclusive of service map, error/thro ...
```
✔️

kuberkaul commented on _posts/2021-12-02-distributed-tracing-pipeline-with-opentelemetry.md:
```diff
@@ -0,0 +1,904 @@
+
+### Sink : OpenSearch
+
+As you may have noticed in the architecture, we utilize OpenSearch to ship to and store traces. [Opensearch](https://aws.amazon.com/blogs/opensource/introducing-opensearch/) makes an excellent choice for storing and searching trace data, along with other observability data due to its fast search capabilities and horizontal scalability.
```
✔️

kuberkaul commented on _posts/2021-12-02-distributed-tracing-pipeline-with-opentelemetry.md:
```diff
@@ -0,0 +1,904 @@
+
+**&#8594;** Trace Analytics plugin for OpenSearch Dashboards and Jaeger visualizes trace data from OpenSearch and allows querying on it.
+
+![tracing_step4](/assets/media/blog-images/2021-12-02-distributed-tracing-pipeline-with-opentelemetry/tracing_step4.png){: .img-fluid}
```
I see what you are saying about the image. I do think the last image with all the components has added verbosity which makes its slightly hard to read. Since I do mention it in the text, I think I can reduce the verbosity in the image to make it more clear/legible. I hope that works as I do like the images in 3D, Ill try to edit the last image and make it less verbose.

dlvenable replied:
I do think that removing some of the text on the labels would help. Some have text in parenthesis which is probably not necessary.

Also, can you change the perspective so that it is not as turned? It looks like it is almost a 45 degree angle. Maybe a 20-25 degree angle would help as well? I'm unsure of how your diagram tool works in this regard.

kuberkaul replied:
So I removed some of the text and resized the image on step 4, it looks much easier to read. Also changed OpenTelemetry Collector to OTEL Collector. Unfortunately with cloudcraft.io, it doesn't allow me to set an angle for the 3D images so this is just what comes by default. 

If this looks good, then I can modify other images as well, to go from OpenTelemetry Collector to OTEL Collector

kuberkaul replied:
> This {: .img-fluid} is rendering in the final HTML as displayed in GitHub. I believe this happens with all the images. I'm unsure how this renders in the final tooling though.

I'm unsure of this as well, @stockholmux might have a better idea since it was his suggestion.

dlvenable replied:
I've only looked in GitHub's rendered markdown using your repo:
https://github.com/kuberkaul/project-website/blob/main/_posts/2021-12-02-distributed-tracing-pipeline-with-opentelemetry.md

I didn't run Jekyll. It might not show up in the HTML which Jekyll renders.

dlvenable replied:
I took a look. I think it does look a lot better. If you change the other images, then they should all be more legible. Thanks!

kuberkaul replied:
awesome and done ✔️

stockholmux replied:
The `{:img-fluid}` makes the images resize as the page scales. It shows up in github because of a different flavour of markdown. 

FYI - Spinning up an instance of the website is pretty quick it might help root out these little issues - that's how I review it.

stockholmux commented:
Major progress! 

The other thing I'm seeing in context with the template is the iconography on the images - on this site we should use the OpenSearch logo and name (just OpenSearch, no need for "AWS" or "Amazon") as well as avoiding and not use any non-OpenSearch product names (Kibana, Elasticsearch namely) 

Looks like the "click to expand" isn't quite working. Let me see if I can figure out why the template is barfing on those. 

I'll do a more detailed review shortly.

Here is how it renders currently: 

![localhost_4000_blog_technical-post_2021_12_distributed-tracing-pipeline-with-opentelemetry_](https://user-images.githubusercontent.com/1152927/144916319-351082ce-d2c0-4df1-9b72-94be0de37054.png)

kuberkaul commented:
>The other thing I'm seeing in context with the template is the iconography on the images - on this site we should use the OpenSearch logo and name (just OpenSearch, no need for "AWS" or "Amazon") as well as avoiding and not use any non-OpenSearch product names (Kibana, Elasticsearch namely)

Gotcha, Ill work on changing the iconography for AWS elasticsearch to opensearch and removing terms like AWS, Amazon and Kibana from the image descriptions as well.

kuberkaul commented:
- Images modified to add opensearch logo from https://opensearch.org/brand.html , removed Kibana, AWS from images
- Modified Data Prepper endpoint from YAML to go from amazon/ to opensearch/ 
- Modified hyperlink to go to https://opensearch.org/ instead of https://aws.amazon.com/blogs/big-data/getting-started-with-trace-analytics-in-amazon-elasticsearch-service/

stockholmux commented:
@kuberkaul I did a PR to your fork with some changes - basically I broke out your "expanding" yaml examples into separate files and converted the HTML into an include.

kuberkaul commented:
@stockholmux Thanks for that change, I merged it. Also added small fixes based on the feedback from Rajiv. I believe you are cc'ed on it. 

He does have feedback on changing **distributed tracing "pipeline"** to **distributed tracing "solution" or "view"** in the title and body. I am okay with the word **_pipeline_** but if you guys feel s**_olution or view (or an alternative)_** is a better fit, I would be happy to make that change through out the document. Let me know and I can do it.

stockholmux commented on _posts/2021-12-02-distributed-tracing-pipeline-with-opentelemetry.md:
```diff
@@ -0,0 +1,248 @@
+
+Since there is no blocklist for filters, we will be using a workaround to sample all but cherrypick services that we do not want sampled. we will be using the `string_attribute` filter to create a negative filter. The filter will be dropping specific service traces that we do not want sampled and sample all other traces. This can be done using the following:
+
+- **&#8594;** In the agent config, add a span attribute for all traces with a key/value pair (e.g - retain_span/false) using attribute processor.
```
drop the leading `-` on the items in this list - it causes a double bullet:

<img width="896" alt="Screen Shot 2021-12-07 at 10 00 39 AM" src="https://user-images.githubusercontent.com/1152927/145073167-4dbb6354-f751-43ea-891e-7adbf2a61862.png">

In the architecture section, the arrows work nicely, so follow that formatting pattern
<img width="900" alt="Screen Shot 2021-12-07 at 10 01 21 AM" src="https://user-images.githubusercontent.com/1152927/145073262-b4764262-8a08-4cbd-bdce-ef849fd4f3d8.png">

kuberkaul replied:
✔️

stockholmux reviewed (changes requested):
I'm so excited for this blog! I had a few little changes - just formatting and grammar mostly. After those are fixed, I'm satisfied.

kuberkaul commented on _posts/2021-12-02-distributed-tracing-pipeline-with-opentelemetry.md:
```diff
@@ -0,0 +1,248 @@
+- [W3C Correlation-Context HTTP Propagator](https://w3c.github.io/correlation-context/)
+- [B3 Zipkin HTTP Propagator](https://github.com/openzipkin/b3-propagation)
+
+This works well for our service mesh PaaS as Istio leverages Envoy’s distributed tracing feature to provide tracing integration out of the box. Specifically, Istio provides options to install various tracing backend and configure proxies to send trace spans to them automatically. It requires an application to propagate the [B3 Zipkin HTTP Propagator](https://github.com/openzipkin/b3-propagation)  ...
```
✔️

kuberkaul commented on _posts/2021-12-02-distributed-tracing-pipeline-with-opentelemetry.md:
```diff
@@ -0,0 +1,248 @@
+
+- **Jaeger UI (query)**
+
+To visualize your traces with Jaeger we need to run a Jaeger query. jaeger-query serves the API endpoints and a React/Javascript UI. The service is stateless and is typically run behind a load balancer:
```
yup its a component of jaeger so Ill capitalize it.

kuberkaul replied:
✔️

kuberkaul commented on _posts/2021-12-02-distributed-tracing-pipeline-with-opentelemetry.md:
```diff
@@ -0,0 +1,248 @@
+  description: "Over the past few years, the importance of observability when developing and managing applications has spiked with the spotlight firmly on micro-services, service mesh. Distributed services can be unpredictable. Despite our best efforts, failures and performance bottlenecks in such systems are inevitable and difficult to isolate. In such an environment, having deep visibility into ...
+---
+
+# Building a distributed tracing pipeline with OpenTelemetry collector, Data Prepper, and OpenSearch trace analytics
```
✔️

kuberkaul commented on _posts/2021-12-02-distributed-tracing-pipeline-with-opentelemetry.md:
```diff
@@ -0,0 +1,248 @@
+
+We will be deploying OpenTelemetry agents as daemonset to receive, process, and export trace from every EKS worker node. The Agent is capable of receiving telemetry data (push and pull based) as well as enhancing telemetry data with metadata such as custom tags or infrastructure information. In addition, the agent can offload responsibilities that client instrumentation would otherwise need to ha ...
+
+The agent can be deployed either as a daemonset or as a sidecar in a Kubernetes (k8s) cluster. This step may be skipped (not recommended), if you are creating this pipeline in a test environment and would rather send traces straight to the open telemetry collector, which is deployed as a kubernetes deployment (horizontally scalable).
```
✔️

kuberkaul commented on _posts/2021-12-02-distributed-tracing-pipeline-with-opentelemetry.md:
```diff
@@ -0,0 +1,248 @@
+
+### Step 3 : OpenTelemetry Gateway
+
+The OpenTelemetry agent forwards the telemetry data to an OpenTelemetry collector gateway. A Gateway cluster runs as a standalone service and offers advanced capabilities over the agent, including tail-based sampling. In addition, a Gateway cluster can limit the number of egress points required to send data, as well as consolidate API token management. Each collector instance in a Gateway cluster ...
```
✔️

kuberkaul commented on _posts/2021-12-02-distributed-tracing-pipeline-with-opentelemetry.md:
````diff
@@ -0,0 +1,248 @@
+          fieldPath: status.hostIP
+```
+
+To setup the pipeline, you can apply the kubernetes manifest files and helm chart(for jaeger). Besides being in the body of this blog, the manifests have been added to [opensearch-blog github repo](https://github.com/dowjones/opensearch-blog/tree/master/source) for convenient access.
````
✔️

kuberkaul commented on _posts/2021-12-02-distributed-tracing-pipeline-with-opentelemetry.md:
```diff
@@ -0,0 +1,248 @@
+
+Once you reach this point, the difficult part is over. You should have distributed traces created, context propagated, and sampled using tail based sampling, transformed, and exported to OpenSearch under different index patterns.
+
+Now is the part where you actually get to use the pipeline and visualize these traces to run queries, build dashboards, setup alerting. While Trace Analytics OpenSearch Dashboards plugin is a managed plugin that comes with OpenSearch, we host our own [Jaeger Frontend/UI](https://www.jaegertracing.io/docs/1.28/frontend-ui/) with a simple jaeger UI to query for distributed traces with Traces View
```
✔️

kuberkaul commented on _posts/2021-12-02-distributed-tracing-pipeline-with-opentelemetry.md:
```diff
@@ -0,0 +1,248 @@
+
+So the next step would naturally be to extend this pipeline to be more than just a distributed tracing pipeline and morph to a distributed telemetry pipeline!
+
+_Comments, questions, corrections? Just let me know on [twitter](https://twitter.com/kub3rkaul), via email kuber.kaul@dowjones.com, or submit a github issue_
```
✔️

kuberkaul commented on _posts/2021-12-02-distributed-tracing-pipeline-with-opentelemetry.md:
```diff
@@ -0,0 +1,248 @@
+
+OpenSearch is a community-driven, open source search and analytics suite derived from Apache 2.0 licensed Elasticsearch 7.10.2 & Kibana 7.10.2. Additionally, it comes with managed Trace Analytics plugin, a visualization and user interface, and OpenSearch Dashboards.
+
+- **Data Prepper** and **Jaeger** to re-format OpenTelemetry trace data and export it to OpenSearch in formats that both the OpenSearch dashboard and Jaeger understand. (This should go away soon as both Jaeger and OpenSearch Dashboards will natively be able to use OTLP traces)
```
✔️

kuberkaul commented on _posts/2021-12-02-distributed-tracing-pipeline-with-opentelemetry.md:
```diff
@@ -0,0 +1,248 @@
+
+#### Creating traces
+
+We need to create/propagate traces to be able to use a distributed tracing pipeline. OpenTelemetry provides a collection of tools such as API, SDK and integrates with popular languages and frameworks to integrate with the greater OpenTelemetry ecosystem, such as OpenTelemetry Protocol (OTLP) and the OpenTelemetry collector.
```
✔️

kuberkaul commented:
@stockholmux I think I got all the changes, you may want to do a final review. I am excited for this to be published and it has been an enjoyable process!

stockholmux commented:
@kuberkaul I'm going to merge this - I see a tiny issue (spelling) but I'll fix it before it gets to production.

stockholmux reviewed (approved):
LGTM

kuberkaul commented:
🎉
